lib/jsonrs.ex

```diff
@@ -106,4 +111,12 @@ defmodule Jsonrs do
   """
   @spec encode_to_iodata!(term, keyword) :: String.t()
   def encode_to_iodata!(input, opts \\ []), do: encode!(input, opts)
+
+  defp validate_compression(nil), do: {:none, 0}
+  defp validate_compression(:gzip), do: {:gzip, nil}
+  defp validate_compression(true), do: {:gzip, nil}
+  defp validate_compression({:gzip, level}) when level in (0..10), do: {:gzip, level}
+  defp validate_compression(other) do
+    raise "invalid :compress option. #{inspect(other)} is not supported."
+  end
```

Comment:
As much as possible, I'd like to rely on the NIF for handling parsing and validating, since it can likely do it much faster. Just for the sake of your intuition, the way Rustler will handle parsing an option in an `Optional<>` is if it can coerce what's passed in into the contained type, it'll do so, otherwise it'll parse as `None`. So things like `nil` will already be handled and treated equivalently to `{:none, 0}` without us needing to map them at the Elixir layer. Likewise, compression levels above 9 will be treated as 9 by the NIF, which I think is acceptable behavior and shouldn't be blocked, especially since later compression methods may support levels above 9. I think pretty much the only mapping we want to do at the Elixir is turn atoms (of any value) into `{atom, nil}` tuples, and let the NIF raise an argument error if the atom isn't one of the supported compression algs, which it already does.

Reply:
I've simplified this and let most of the argument validation/coercion happen in rust. I tend to do as much as I can do efficiently within Elixir for maintainability, but totally get your preference here so I've updated to only handle the `atom => {atom, nil}` use-case. I also added several test cases with passing invalid arguments to make sure we document the expected behavior

Reply:
hm, I'm trying to think of a good way to do the `atom -> {atom, nil}` conversion without also applying it to `nil` and `false` inputs, which I think should simply not enable compression rather than raise.

Also just to clarify, I'm nitpicking this because it runs on every invocation of encode whether or not the user is calling for compression. If the user opts into compression and something isn't maximally performant, that's okay, but if they're just using Jsonrs without compression, the existence of the compression feature should affect them as little as possible. I think I'd like to even avoid the function call and just have an inline case or if statement like when handling the `indent` and `lean` options, but I don't even know for sure that that would make a measurable difference.

Reply:
I've never had any performance problems with breaking out a private function in an elixir module. This is because the [erlang compiler](https://www.erlang.org/docs/20/man/compile.html) will do function inlining within a module:

> The compiler can do function inlining within an Erlang module. Inlining means that a call to a function is replaced with the function body with the arguments replaced with the actual values. The semantics are preserved, except if exceptions are generated in the inlined code

I wrote a little benchmark to validate my understanding and I found that an inline `case` has the same performance as a helper function, and both of them only add about 1/10th of a microsecond for the caller.

<details>
  <summary>Benchmark and Results</summary>

```option_parsing.exs
# Sample file generated by http://json-generator.com/
# stored as gzip to minimize the git repo impact
sample = File.read!("bench/fixtures/1kb_sample.json.gz") |> :zlib.gunzip() |> Jason.decode!()

defmodule Options do
  def encode_inline!(input, opts) do
    {lean, opts} = Keyword.pop(opts, :lean, false)
    {indent, opts} = Keyword.pop(opts, :pretty, nil)
    {compression, _opts} = Keyword.pop(opts, :compress, :none)
    compression = case compression do
      false -> :none
      nil -> :none
      atom when is_atom(atom) -> {atom, nil}
      other -> other
    end
    indent = if true == indent, do: 2, else: indent
    case lean do
      true -> input
      false -> Jsonrs.Encoder.encode(input)
    end
    |> Jsonrs.nif_encode(indent, compression)
  end

  def encode_helper!(input, opts) do
    {lean, opts} = Keyword.pop(opts, :lean, false)
    {indent, opts} = Keyword.pop(opts, :pretty, nil)
    {compression, _opts} = Keyword.pop(opts, :compress, :none)
    compression = validate_compression(compression)
    indent = if true == indent, do: 2, else: indent
    case lean do
      true -> input
      false -> Jsonrs.Encoder.encode(input)
    end
    |> Jsonrs.nif_encode(indent, compression)
  end

  def encode_no_check!(input, opts) do
    {lean, opts} = Keyword.pop(opts, :lean, false)
    {indent, _opts} = Keyword.pop(opts, :pretty, nil)
    indent = if true == indent, do: 2, else: indent
    case lean do
      true -> input
      false -> Jsonrs.Encoder.encode(input)
    end
    |> Jsonrs.nif_encode(indent, {:none, nil})
  end

  defp validate_compression(nil), do: :none
  defp validate_compression(false), do: :none
  defp validate_compression(atom) when is_atom(atom), do: {atom, nil}
  defp validate_compression(other), do: other
end

Benchee.run(
  %{
    "encode_inline" => fn -> Options.encode_inline!(sample, lean: true) end,
    "encode_helper" => fn -> Options.encode_helper!(sample, lean: true) end,
    "encode_no_check" => fn -> Options.encode_no_check!(sample, lean: true) end,
  },
  time: 20,
  memory_time: 2
)
```

```
Operating System: macOS
CPU Information: Apple M1
Number of Available Cores: 8
Available memory: 16 GB
Elixir 1.14.3
Erlang 25.2.3

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 20 s
memory time: 2 s
reduction time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 1.20 min

Benchmarking encode_helper ...
Benchmarking encode_inline ...
Benchmarking encode_no_check ...

Name                      ips        average  deviation         median         99th %
encode_helper        102.45 K        9.76 μs   ±176.45%        8.08 μs       36.96 μs
encode_inline        100.14 K        9.99 μs   ±171.05%        8.17 μs       37.33 μs
encode_no_check       98.34 K       10.17 μs   ±162.94%        8.17 μs       37.50 μs

Comparison: 
encode_helper        102.45 K
encode_inline        100.14 K - 1.02x slower +0.23 μs
encode_no_check       98.34 K - 1.04x slower +0.41 μs

Memory usage statistics:

Name               Memory usage
encode_helper             168 B
encode_inline             168 B - 1.00x memory usage +0 B
encode_no_check           120 B - 0.71x memory usage -48 B

**All measurements for memory usage were the same**
```
</details>